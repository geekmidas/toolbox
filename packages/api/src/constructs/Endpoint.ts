import type { StandardSchemaV1 } from '@standard-schema/spec';
import set from 'lodash.set';
import type { OpenAPIV3_1 } from 'openapi-types';
import { UnprocessableEntityError } from '../errors';
import type { Logger } from '../logger';

import type {
  Service,
  ServiceDiscovery,
  ServiceRecord,
} from '../service-discovery';
import {
  Function,
  type FunctionContext,
  type FunctionHandler,
} from './Function';
import { convertStandardSchemaToJsonSchema } from './helpers';
import { type OpenApiSchemaOptions, buildOpenApiSchema } from './openapi';
import {
  FunctionType,
  type HttpMethod,
  type InferComposableStandardSchema,
  type InferStandardSchema,
  type LowerHttpMethod,
  type RemoveUndefined,
} from './types';

export class Endpoint<
  TRoute extends string,
  TMethod extends HttpMethod,
  TInput extends EndpointSchemas = {},
  OutSchema extends StandardSchemaV1 | undefined = undefined,
  TServices extends Service[] = [],
  TLogger extends Logger = Logger,
  TSession = unknown,
> extends Function<TInput, TServices, TLogger, OutSchema> {
  route: TRoute;
  method: TMethod;
  description?: string;
  public readonly status: SuccessStatus;
  public getSession: SessionFn<TServices, TLogger, TSession> = () =>
    ({}) as TSession;
  public authorize: AuthorizeFn<TServices, TLogger, TSession> = () => true;

  static async buildOpenApiSchema(
    endpoints: Endpoint<any, any, any, any, any, any>[],
    options?: OpenApiSchemaOptions,
  ) {
    return buildOpenApiSchema(endpoints, options);
  }

  static validate<T extends StandardSchemaV1>(schema: T, data: unknown) {
    return schema['~standard'].validate(data);
  }

  get fullPath() {
    return `${this.method} ${this._path}` as const;
  }

  static async parseSchema<T extends StandardSchemaV1>(
    schema: T,
    data: unknown,
  ): Promise<InferStandardSchema<T>> {
    if (!schema) {
      return undefined as InferStandardSchema<T>;
    }

    const parsed = await Endpoint.validate(
      schema as unknown as StandardSchemaV1,
      data,
    );
    if (parsed.issues) {
      throw new UnprocessableEntityError('Validation failed', parsed.issues);
    }

    return parsed.value as InferStandardSchema<T>;
  }

  async parseOutput(output: unknown): Promise<InferStandardSchema<OutSchema>> {
    return Endpoint.parseSchema(
      this.outputSchema as StandardSchemaV1,
      output,
    ) as Promise<InferStandardSchema<OutSchema>>;
  }

  async parseInput<K extends keyof TInput>(
    input: unknown,
    key: K,
  ): Promise<InferComposableStandardSchema<TInput[K]>> {
    const schema = this.input?.[key];
    return Endpoint.parseSchema(schema as StandardSchemaV1, input) as Promise<
      InferComposableStandardSchema<TInput[K]>
    >;
  }

  async parseBody(body: unknown): Promise<InferStandardSchema<TInput['body']>> {
    return this.parseInput(body, 'body') as Promise<
      InferStandardSchema<TInput['body']>
    >;
  }

  static createHeaders(headers: Record<string, string>) {
    const headerMap = new Map<string, string>();
    for (const [k, v] of Object.entries(headers)) {
      const key = k.toLowerCase();
      headerMap.set(key, v);
    }

    return function get(key: string): string | undefined {
      return headerMap.get(key.toLowerCase());
    };
  }

  handler: EndpointHandler<TInput, TServices, TLogger, OutSchema, TSession> = (
    ctx: EndpointContext<TInput, TServices, TLogger, TSession>,
  ): OutSchema extends StandardSchemaV1
    ? InferStandardSchema<OutSchema> | Promise<InferStandardSchema<OutSchema>>
    : void | Promise<void> => {
    return this.fn({
      body: ctx.body,
      query: ctx.query,
      params: ctx.params,
      services: ctx.services,
      logger: ctx.logger,
      header: ctx.header,
      session: ctx.session,
    } as unknown as FunctionContext<TInput, TServices, TLogger>);
  };

  static isEndpoint(obj: any): obj is Endpoint<any, any, any, any> {
    return (
      obj &&
      (obj as Function).__IS_FUNCTION__ === true &&
      obj.type === FunctionType.Endpoint
    );
  }

  get _path() {
    return this.route.replace(/:(\w+)/g, '{$1}') as ConvertRouteParams<TRoute>;
  }

  async toOpenApi3Route(): Promise<EndpointOpenApiSchema<TRoute, TMethod>> {
    const operation: OpenAPIV3_1.OperationObject = {
      ...(this.description && { description: this.description }),
      responses: {
        '200': {
          description: 'Successful response',
        } as OpenAPIV3_1.ResponseObject,
      },
    };

    // Add response schema
    if (this.outputSchema) {
      const responseSchema = await convertStandardSchemaToJsonSchema(
        this.outputSchema,
      );
      if (responseSchema) {
        set(
          operation,
          ['responses', '200', 'content', 'application/json', 'schema'],
          responseSchema,
        );
      }
    }

    // Add parameters array
    const parameters: OpenAPIV3_1.ParameterObject[] = [];

    // Since the EndpointBuilder doesn't have body/search/params methods yet,
    // and the input is a composite type, we need to check if input exists
    // and has the expected shape
    if (this.input && typeof this.input === 'object') {
      // Add request body for methods that support it
      if (
        ['POST', 'PUT', 'PATCH'].includes(this.method) &&
        'body' in this.input &&
        this.input.body
      ) {
        const bodySchema = await convertStandardSchemaToJsonSchema(
          this.input.body as StandardSchemaV1,
        );
        if (bodySchema) {
          set(operation, ['requestBody'], {
            required: true,
            content: {
              'application/json': {
                schema: bodySchema,
              },
            },
          });
        }
      }

      // Add path parameters
      if ('params' in this.input && this.input.params) {
        const paramsSchema = await convertStandardSchemaToJsonSchema(
          this.input.params as StandardSchemaV1,
        );
        if (
          paramsSchema &&
          paramsSchema.type === 'object' &&
          paramsSchema.properties
        ) {
          for (const [name, schema] of Object.entries(
            paramsSchema.properties,
          )) {
            parameters.push({
              name,
              in: 'path',
              required: paramsSchema.required?.includes(name) ?? true,
              schema: schema as any,
            });
          }
        }
      }

      // Add query parameters
      if ('query' in this.input && this.input.query) {
        const querySchema = await convertStandardSchemaToJsonSchema(
          this.input.query,
        );
        if (
          querySchema &&
          querySchema.type === 'object' &&
          querySchema.properties
        ) {
          for (const [name, schema] of Object.entries(querySchema.properties)) {
            parameters.push({
              name,
              in: 'query',
              required: querySchema.required?.includes(name) ?? false,
              schema: schema as any,
            });
          }
        }
      }
    }

    if (parameters.length > 0) {
      operation.parameters = parameters;
    }

    return {
      [this._path]: {
        [this.method.toLowerCase()]: operation,
      },
    } as EndpointOpenApiSchema<TRoute, TMethod>;
  }

  constructor({
    fn,
    method,
    route,
    description,
    input,
    logger,
    output: outputSchema,
    services,
    timeout,
    getSession,
    authorize,
    status = SuccessStatus.OK,
  }: EndpointOptions<
    TRoute,
    TMethod,
    TInput,
    OutSchema,
    TServices,
    TLogger,
    TSession
  >) {
    super(
      fn as FunctionHandler<TInput, TServices, TLogger, OutSchema>,
      timeout,
      FunctionType.Endpoint,
      input,
      outputSchema,
      services,
      logger,
    );

    this.route = route;
    this.method = method;
    this.description = description;
    this.status = status;
    if (getSession) {
      this.getSession = getSession;
    }

    if (authorize) {
      this.authorize = authorize;
    }
  }
}

export type EndpointInput<
  TBody extends StandardSchemaV1 | undefined = undefined,
  TSearch extends StandardSchemaV1 | undefined = undefined,
  TParams extends StandardSchemaV1 | undefined = undefined,
> = RemoveUndefined<{
  body: TBody;
  search: TSearch;
  params: TParams;
}>;

export interface EndpointOptions<
  TRoute extends string,
  TMethod extends HttpMethod,
  TInput extends EndpointSchemas = {},
  TOutput extends StandardSchemaV1 | undefined = undefined,
  TServices extends Service[] = [],
  TLogger extends Logger = Logger,
  TSession = unknown,
> {
  route: TRoute;
  method: TMethod;
  fn: EndpointHandler<TInput, TServices, TLogger, TOutput, TSession>;
  authorize: AuthorizeFn<TServices, TLogger, TSession> | undefined;
  description: string | undefined;
  timeout: number | undefined;
  input: TInput | undefined;
  output: TOutput | undefined;
  services: TServices;
  logger: TLogger;
  getSession: SessionFn<TServices, TLogger, TSession> | undefined;
  status: SuccessStatus | undefined;
}

export type EndpointSchemas = Partial<{
  params: StandardSchemaV1;
  query: StandardSchemaV1;
  body: StandardSchemaV1;
}>;

export type AuthorizeFn<
  TServices extends Service[] = [],
  TLogger extends Logger = Logger,
  TSession = unknown,
> = (
  ctx: FunctionContext<{}, TServices, TLogger> & {
    header: HeaderFn;
    session: TSession;
  },
) => Promise<boolean> | boolean;

export type SessionFn<
  TServices extends Service[] = [],
  TLogger extends Logger = Logger,
  TSession = unknown,
> = (
  ctx: FunctionContext<{}, TServices, TLogger> & { header: HeaderFn },
) => Promise<TSession> | TSession;

export type ConvertRouteParams<T extends string> =
  T extends `${infer Start}:${infer Param}/${infer Rest}`
    ? `${Start}{${Param}}/${ConvertRouteParams<Rest>}`
    : T extends `${infer Start}:${infer Param}`
      ? `${Start}{${Param}}`
      : T;

export type EndpointOpenApiSchema<
  TRoute extends string,
  TMethod extends HttpMethod,
> = {
  [key in ConvertRouteParams<TRoute>]: {
    [key in LowerHttpMethod<TMethod>]: OpenAPIV3_1.OperationObject<{}>;
  };
};

export type EndpointHeaders = Map<string, string>;
export type HeaderFn = (key: string) => string | undefined;

export type EndpointContext<
  Input extends EndpointSchemas | undefined = undefined,
  TServices extends Service[] = [],
  TLogger extends Logger = Logger,
  TSession = unknown,
> = {
  services: ServiceDiscovery<ServiceRecord<TServices>>;
  logger: TLogger;
  header: HeaderFn;
  session: TSession;
} & InferComposableStandardSchema<Input>;

export type EndpointHandler<
  TInput extends EndpointSchemas | undefined = undefined,
  TServices extends Service[] = [],
  TLogger extends Logger = Logger,
  OutSchema extends StandardSchemaV1 | undefined = undefined,
  TSession = unknown,
> = (
  ctx: EndpointContext<TInput, TServices, TLogger, TSession>,
) => OutSchema extends StandardSchemaV1
  ? InferStandardSchema<OutSchema> | Promise<InferStandardSchema<OutSchema>>
  : any | Promise<any>;

export enum SuccessStatus {
  OK = 200,
  Created = 201,
  Accepted = 202,
  NoContent = 204,
  ResetContent = 205,
  PartialContent = 206,
}
