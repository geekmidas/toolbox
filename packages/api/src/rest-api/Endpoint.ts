import type { StandardSchemaV1 } from '@standard-schema/spec';
import uniqBy from 'lodash.uniqby';

import { UnprocessableEntityError } from '../errors';
import { ConsoleLogger, type Logger } from '../logger';
import type { HermodServiceConstructor } from '../services';
import type {
  Authorizer,
  ConvertRouteParams,
  EndpointSchemas,
  HandlerContext,
  InferStandardSchema,
  JoinPaths,
  LowerMethod,
  Method,
  Route,
  RouteHandler,
  SessionFn,
} from './types';

// Add a SuccessStatus enum for common HTTP success codes
export enum SuccessStatus {
  OK = 200,
  Created = 201,
  Accepted = 202,
  NoContent = 204,
  ResetContent = 205,
  PartialContent = 206,
}

const DEFAULT_LOGGER = new ConsoleLogger() as any;
export class Handler<
  S extends EndpointSchemas,
  Path extends string,
  TMethod extends Method,
  OutSchema extends StandardSchemaV1 | undefined = undefined,
  TServices extends HermodServiceConstructor[] = [],
  TLogger extends Logger = ConsoleLogger,
  TSession = unknown,
> {
  __IS_HANDLER__ = true;

  static isHandler(obj: any): obj is Handler<any, any, any, any, any, any> {
    return obj && obj.__IS_HANDLER__ === true;
  }

  constructor(
    public readonly method: TMethod,
    public readonly path: Path,
    public readonly schemas: S,
    public readonly outputSchema: OutSchema,
    public readonly authorize: Authorizer<S, TServices, TLogger, TSession>,
    public readonly handler: RouteHandler<
      S,
      OutSchema,
      TServices,
      TLogger,
      TSession
    >,
    public readonly status: SuccessStatus = SuccessStatus.OK,
    private description?: string,
    public services: TServices = {} as TServices,
    public logger: TLogger = DEFAULT_LOGGER,
    public getSession: SessionFn<S, TServices, TLogger, TSession> = () =>
      ({}) as TSession,
  ) {}

  get _path() {
    return this.path.replace(/:(\w+)/g, '{$1}') as ConvertRouteParams<Path>;
  }

  route(): Route<TMethod, ConvertRouteParams<Path>> {
    return `${this.method} ${this._path}`;
  }

  static StandardSchemaJsonSchema = {
    zod: async (schema) => {
      const { z } = await import('zod/v4');
      return z.toJSONSchema(schema, { unrepresentable: 'any' });
    },
    valibot: async (schema) => {
      const { toJsonSchema } = await import('@valibot/to-json-schema');
      return toJsonSchema(schema as any);
    },
  };

  async parseSchema<T extends StandardSchemaV1>(
    data?: any,
    schema?: T,
  ): Promise<InferStandardSchema<T> | undefined> {
    if (!schema) {
      return undefined;
    }

    const validated = await schema['~standard'].validate(data);

    if (!validated.issues) {
      return validated.value as InferStandardSchema<T>;
    }

    this.logger.error(validated.issues);

    throw new UnprocessableEntityError(undefined, validated.issues);
  }

  async parseParams(
    data?: any,
  ): Promise<InferStandardSchema<S['params']> | undefined> {
    return this.parseSchema(data, this.schemas.params) as InferStandardSchema<
      S['params']
    >;
  }

  async parseBody(
    data?: any,
  ): Promise<InferStandardSchema<S['body']> | undefined> {
    this.logger.debug({ data }, 'Parsing body schema');
    return this.parseSchema(data, this.schemas.body) as InferStandardSchema<
      S['body']
    >;
  }

  async parseQuery(
    data?: any,
  ): Promise<InferStandardSchema<S['query']> | undefined> {
    return this.parseSchema(data, this.schemas.query) as InferStandardSchema<
      S['query']
    >;
  }

  async parseOutput(
    data?: any,
  ): Promise<InferStandardSchema<OutSchema> | undefined> {
    return this.parseSchema(
      data,
      this.outputSchema,
    ) as InferStandardSchema<OutSchema>;
  }

  async toJSONSchema(schema?: StandardSchemaV1): Promise<any | undefined> {
    if (!schema) {
      return undefined;
    }

    const vendor = schema['~standard']
      .vendor as keyof typeof Handler.StandardSchemaJsonSchema;
    const toJsonSchema = Handler.StandardSchemaJsonSchema[vendor];

    return toJsonSchema?.(schema);
  }

  async toOpenAPI(): Promise<{
    [P in Path]: {
      [M in LowerMethod<TMethod>]: {
        parameters: any[];
        requestBody?: any;
        responses: any;
      };
    };
  }> {
    const method = this.method.toLowerCase() as LowerMethod<typeof this.method>;
    const parameters: any[] = [];
    const responses = {};

    // Convert params schema to OpenAPI parameters
    if (this.schemas.params) {
      const paramsSchema = await this.toJSONSchema(this.schemas.params);
      if (paramsSchema?.properties) {
        for (const [name, schema] of Object.entries(paramsSchema.properties)) {
          // @ts-ignore
          delete schema.$schema;
          parameters.push({
            name,
            in: 'path',
            required: Array.isArray(paramsSchema.required)
              ? paramsSchema.required.includes(name)
              : false,
            // @ts-ignore
            ...schema,
          });
        }
      }
    }

    // Convert query schema to OpenAPI parameters
    if (this.schemas.query) {
      const querySchema = await this.toJSONSchema(this.schemas.query);
      if (querySchema?.properties) {
        for (const [name, schema] of Object.entries(querySchema.properties)) {
          // @ts-ignore
          delete schema.$schema;
          parameters.push({
            name,
            in: 'query',
            required: Array.isArray(querySchema.required)
              ? querySchema.required.includes(name)
              : false,
            // @ts-ignore
            ...schema,
          });
        }
      }
    }

    // Convert body schema to OpenAPI requestBody
    if (this.schemas.body) {
      const schema = await this.toJSONSchema(this.schemas.body);
      // @ts-ignore
      delete schema.$schema;
      parameters.push({
        name: 'body',
        in: 'body',
        required: true,
        schema,
      });
    }

    // Convert output schema to OpenAPI response
    if (this.outputSchema) {
      const outputSchema = await this.toJSONSchema(this.outputSchema);
      delete outputSchema.$schema;
      responses[this.status] = {
        description: this.description || 'Success',
        schema: outputSchema,
      };
    } else {
      responses[this.status] = {
        description: 'Success',
      };
    }

    return {
      [this._path]: {
        [method]: {
          produces: this.outputSchema ? ['application/json'] : undefined,
          parameters,
          responses,
        },
      },
    } as any;
  }
}

class EndpointBuilder<
  TSchema extends EndpointSchemas = {},
  Path extends string = string,
  TMethod extends Method = Method,
  OutSchema extends StandardSchemaV1 | undefined = undefined,
  TServices extends HermodServiceConstructor[] = [],
  TLogger extends Logger = ConsoleLogger,
  TSession = unknown,
> {
  private method: TMethod;
  private path: Path;
  private _description?: string;
  private schemas: EndpointSchemas = {};
  private outputSchema?: OutSchema;
  private statusCode: SuccessStatus = SuccessStatus.OK;

  // Made these accessible to EndpointFactory
  // @ts-ignore
  _services: TServices = [];
  _logger: TLogger = DEFAULT_LOGGER;
  _auth: TSession = {} as TSession;
  authorizeFn: Authorizer<TSchema, TServices, TLogger, TSession> = () => true;
  _getSession: SessionFn<TSchema, TServices, TLogger, TSession> = () =>
    ({}) as TSession;

  constructor(method: TMethod, path: Path) {
    this.method = method;
    this.path = path;
  }

  params<T extends StandardSchemaV1>(
    schema: T,
  ): EndpointBuilder<
    Omit<TSchema, 'params'> & { params: T },
    Path,
    TMethod,
    OutSchema,
    TServices,
    TLogger,
    TSession
  > {
    this.schemas.params = schema;
    // @ts-expect-error - for chaining with updated generic
    return this;
  }

  services<T extends HermodServiceConstructor[]>(
    services: T,
  ): EndpointBuilder<
    TSchema,
    Path,
    TMethod,
    OutSchema,
    [...TServices, ...T],
    TLogger,
    TSession
  > {
    this._services = uniqBy(
      [...this._services, ...services],
      (s) => s.serviceName,
    ) as TServices;
    // @ts-expect-error - for chaining with updated generic
    return this;
  }

  query<T extends StandardSchemaV1>(
    schema: T,
  ): EndpointBuilder<
    Omit<TSchema, 'query'> & { query: T },
    Path,
    TMethod,
    OutSchema,
    TServices,
    TLogger,
    TSession
  > {
    this.schemas.query = schema;
    // @ts-expect-error - for chaining with updated generic
    return this;
  }

  body<T extends StandardSchemaV1>(
    schema: T,
  ): EndpointBuilder<
    Omit<TSchema, 'body'> & { body: T },
    Path,
    TMethod,
    OutSchema,
    TServices,
    TLogger,
    TSession
  > {
    this.schemas.body = schema;
    // @ts-expect-error - for chaining with updated generic
    return this;
  }

  output<T extends StandardSchemaV1>(
    schema: T,
  ): EndpointBuilder<TSchema, Path, TMethod, T, TServices, TLogger, TSession> {
    this.outputSchema = schema as unknown as OutSchema;
    // @ts-expect-error - for chaining with updated generic
    return this;
  }

  status(status: SuccessStatus): this {
    this.statusCode = status;
    return this;
  }

  authorize(
    fn: (
      ctx: HandlerContext<TSchema, TServices, TLogger, TSession>,
    ) => boolean | Promise<boolean>,
  ): this {
    this.authorizeFn = fn;
    return this;
  }

  description(description: string): this {
    this._description = description;
    return this;
  }

  handle(
    fn: (
      ctx: HandlerContext<TSchema, TServices, TLogger, TSession>,
    ) => OutSchema extends StandardSchemaV1
      ? InferStandardSchema<OutSchema> | Promise<InferStandardSchema<OutSchema>>
      : any | Promise<any>,
  ): Handler<TSchema, Path, TMethod, OutSchema, TServices, TLogger, TSession> {
    return new Handler(
      this.method,
      this.path,
      this.schemas as TSchema,
      this.outputSchema,
      this.authorizeFn,
      fn,
      this.statusCode,
      this._description,
      this._services,
      this._logger,
      this._getSession,
    ) as Handler<
      TSchema,
      Path,
      TMethod,
      OutSchema,
      TServices,
      TLogger,
      TSession
    >;
  }
}

// New: EndpointFactory for creating routers and configured endpoints
export class EndpointFactory<
  TServices extends HermodServiceConstructor[] = [],
  TBasePath extends string = '',
  TLogger extends Logger = ConsoleLogger,
  TSession = unknown,
> {
  private defaultServices: TServices;
  constructor(
    defaultServices: TServices,
    private basePath: TBasePath = '' as TBasePath,
    private defaultAuthorizeFn?: (
      ctx: HandlerContext<any, TServices, TLogger, TSession>,
    ) => boolean | Promise<boolean>,
    private defaultLogger: TLogger = DEFAULT_LOGGER,
    private defaultSessionExtractor?: SessionFn<
      {},
      TServices,
      TLogger,
      TSession
    >,
  ) {
    // Initialize default services
    this.defaultServices = uniqBy(
      defaultServices,
      (s) => s.serviceName,
    ) as TServices;
  }

  static joinPaths<TBasePath extends string, P extends string>(
    path: P,
    basePath: TBasePath = '' as TBasePath,
  ): JoinPaths<TBasePath, P> {
    const base = basePath.endsWith('/') ? basePath.slice(0, -1) : basePath;
    const segment = path.startsWith('/') ? path : '/' + path;

    return (base + segment) as JoinPaths<TBasePath, P>;
  }

  // Create a sub-router with a path prefix
  route<TPath extends string>(
    path: TPath,
  ): EndpointFactory<
    TServices,
    JoinPaths<TBasePath, TPath>,
    TLogger,
    TSession
  > {
    const newBasePath = EndpointFactory.joinPaths(path, this.basePath);
    return new EndpointFactory(
      this.defaultServices,
      newBasePath,
      this.defaultAuthorizeFn,
      this.defaultLogger,
      this.defaultSessionExtractor,
    );
  }

  // Create a new factory with authorization
  authorize(
    fn: (
      ctx: HandlerContext<any, TServices, TLogger, TSession>,
    ) => boolean | Promise<boolean>,
  ): EndpointFactory<TServices, TBasePath, TLogger, TSession> {
    return new EndpointFactory(
      this.defaultServices,
      this.basePath,
      fn,
      this.defaultLogger,
      this.defaultSessionExtractor,
    );
  }

  // Create a new factory with services
  services<S extends HermodServiceConstructor[]>(
    services: S,
  ): EndpointFactory<[...S, ...TServices], TBasePath, TLogger, TSession> {
    return new EndpointFactory<
      [...S, ...TServices],
      TBasePath,
      TLogger,
      TSession
    >(
      [...services, ...this.defaultServices],
      this.basePath,
      this.defaultAuthorizeFn,
      this.defaultLogger,
      this.defaultSessionExtractor,
    );
  }

  logger<L extends Logger>(
    logger: L,
  ): EndpointFactory<TServices, TBasePath, L, TSession> {
    return new EndpointFactory<TServices, TBasePath, L, TSession>(
      this.defaultServices,
      this.basePath,
      this.defaultAuthorizeFn as any,
      logger,
      this.defaultSessionExtractor as any,
    );
  }

  session<TSession>(session: SessionFn<{}, TServices, TLogger, TSession>) {
    return new EndpointFactory<TServices, TBasePath, TLogger, TSession>(
      this.defaultServices,
      this.basePath,
      this.defaultAuthorizeFn as any,
      this.defaultLogger,
      session,
    );
  }

  private createBuilder<TMethod extends Method, TPath extends string>(
    method: TMethod,
    path: TPath,
  ): EndpointBuilder<
    {},
    JoinPaths<TBasePath, TPath>,
    TMethod,
    undefined,
    TServices,
    TLogger,
    TSession
  > {
    const fullPath = EndpointFactory.joinPaths(path, this.basePath);
    const builder = new EndpointBuilder(method, fullPath);

    if (this.defaultAuthorizeFn) {
      // @ts-ignore
      builder.authorizeFn = this.defaultAuthorizeFn;
    }
    if (this.defaultServices.length) {
      // @ts-ignore
      builder._services = this.defaultServices as TServices;
    }

    if (this.defaultLogger) {
      // @ts-ignore
      builder._logger = this.defaultLogger as TLogger;
    }

    if (this.defaultSessionExtractor) {
      // @ts-ignore
      builder._getSession = this.defaultSessionExtractor as SessionFn<
        {},
        TServices,
        TLogger,
        TSession
      >;
    }

    return builder as unknown as EndpointBuilder<
      {},
      JoinPaths<TBasePath, TPath>,
      TMethod,
      undefined,
      TServices,
      TLogger,
      TSession
    >;
  }

  post<TPath extends string>(path: TPath) {
    return this.createBuilder('POST', path);
  }

  get<TPath extends string>(path: TPath) {
    return this.createBuilder('GET', path);
  }

  put<TPath extends string>(path: TPath) {
    return this.createBuilder('PUT', path);
  }

  delete<TPath extends string>(path: TPath) {
    return this.createBuilder('DELETE', path);
  }

  patch<TPath extends string>(path: TPath) {
    return this.createBuilder('PATCH', path);
  }

  options<TPath extends string>(path: TPath) {
    return this.createBuilder('OPTIONS', path);
  }
}

// Original Endpoint class for backward compatibility
export class Endpoint {
  static post<TPath extends string>(path: TPath) {
    return new EndpointBuilder('POST', path);
  }
  static get<TPath extends string>(path: TPath) {
    return new EndpointBuilder('GET', path);
  }
  static put<TPath extends string>(path: TPath) {
    return new EndpointBuilder('PUT', path);
  }
  static delete<TPath extends string>(path: TPath) {
    return new EndpointBuilder('DELETE', path);
  }
  static patch<TPath extends string>(path: TPath) {
    return new EndpointBuilder('PATCH', path);
  }
  static options<TPath extends string>(path: TPath) {
    return new EndpointBuilder('OPTIONS', path);
  }

  // New static methods for factory pattern
  static route<P extends string>(
    path: P,
  ): EndpointFactory<HermodServiceConstructor[], P> {
    return new EndpointFactory([], path);
  }

  static authorize(
    fn: (
      ctx: HandlerContext<any, HermodServiceConstructor[]>,
    ) => boolean | Promise<boolean>,
  ): EndpointFactory<HermodServiceConstructor[]> {
    return new EndpointFactory([], '', fn);
  }

  static services<S extends HermodServiceConstructor[]>(
    services: S,
  ): EndpointFactory<S> {
    return new EndpointFactory(services);
  }
}

// Export a default instance for convenience
export const e = new EndpointFactory([]);

// Utility function to create a new factory
export function createEndpoint() {
  return new EndpointFactory([]);
}
