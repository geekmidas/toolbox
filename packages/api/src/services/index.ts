/**
 * BACKWARD COMPATIBILITY RE-EXPORTS
 *
 * Services have been moved to '@geekmidas/services' package.
 *
 * @deprecated Import from '@geekmidas/services' instead.
 * These re-exports will be removed in v3.0.0.
 */

export type {
  Service,
  ServiceRecord,
  ExtractServiceNames,
  HermodServiceInterface,
} from '@geekmidas/services';

export { ServiceDiscovery } from '@geekmidas/services';
