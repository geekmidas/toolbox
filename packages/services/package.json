{
  "name": "@geekmidas/services",
  "version": "0.0.1",
  "private": false,
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.mjs",
      "require": "./dist/index.cjs"
    }
  },
  "scripts": {
    "ts": "tsc --noEmit --skipLibCheck src/**/*.ts"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/geekmidas/toolbox.git"
  },
  "publishConfig": {
    "registry": "https://registry.npmjs.org/",
    "access": "public"
  },
  "dependencies": {
    "@geekmidas/logger": "workspace:*"
  },
  "peerDependencies": {
    "@geekmidas/envkit": "workspace:*"
  }
}
