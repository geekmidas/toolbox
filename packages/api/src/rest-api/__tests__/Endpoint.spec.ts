import { describe, expect, it } from 'vitest';
import { z } from 'zod/v4';
import { HermodService } from '../../services';
import { e } from '../Endpoint';

class TestService extends HermodService<{
  name: string;
}> {
  static readonly serviceName = 'TestService';
  register() {
    return Promise.resolve({
      name: 'TestService',
    });
  }
}

class TestService2 extends HermodService<{}> {
  static readonly serviceName = 'TestService2';
  register(): Promise<{}> {
    return Promise.resolve({});
  }
}

const E = e.services([TestService]);

describe('Endpoint', () => {
  it('Endpoint.toJSONSchema', async () => {
    const getUsers = E.get('/users')
      .query(z.object({ q: z.string().min(2).max(100) }))
      .description('Get users')
      .output(
        z.object({
          name: z.string(),
        }),
      )
      .handle(({ services }) => ({
        name: '',
      }));

    const schema = await getUsers.toOpenAPI();

    expect(schema).toHaveProperty('/users');
    expect(schema['/users']).toHaveProperty('get');
  });

  it('Endpoint.toJSONSchema PUT', async () => {
    const updateUsers = E.put('/users')
      .body(z.object({ name: z.string().min(2).max(100) }))
      .description('Get users')
      .output(
        z.object({
          name: z.string(),
        }),
      )
      .handle(({ services }) => {
        return {
          name: '',
        };
      });

    const schema = await updateUsers.toOpenAPI();

    expect(schema).toHaveProperty('/users');
    expect(schema['/users']).toHaveProperty('put');
  });

  it('Should have services', async () => {
    const createUser = E.post('/users')
      .body(z.object({ name: z.string().min(2).max(100) }))
      .services([TestService2])
      .description('Get users')
      .output(
        z.object({
          name: z.string(),
        }),
      )
      .handle(({ services }) => {
        return {
          name: '',
        };
      });

    expect(createUser.services).toHaveLength(2);
  });
});
