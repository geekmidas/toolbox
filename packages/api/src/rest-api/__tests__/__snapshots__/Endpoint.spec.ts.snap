// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Endpoint > Should have services 1`] = `
[
  "TestService",
  "TestService2",
]
`;
